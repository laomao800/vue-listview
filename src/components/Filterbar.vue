<template>
  <el-form
    :inline="true"
    size="small"
    class="lv__filterbar"
    @submit.native.prevent
    @keydown.native.enter="handleSubmit"
  >
    <div v-if="$slots['filterbar-top']" class="lv__filterbar-top">
      <slot name="filterbar-top" />
    </div>

    <div class="lv__filterbar-main">
      <div v-if="$slots['filterbar-left']" class="lv__filterbar-left">
        <slot name="filterbar-left" />
      </div>

      <div
        v-if="isShowFilterButtons || isShowFilterSubmit || isShowFilterFields"
        :class="[
          'lv__filterbar-inner',
          { 'lv__filterbar-inner--fold': isFold },
        ]"
      >
        <!-- 提交、重置按钮区域 -->
        <div
          v-if="isShowFilterSubmit"
          ref="action"
          :class="[
            'lv__filterbar-action',
            {
              'lv__filterbar-action--nomore': isNoMore,
              'lv__filterbar-action--onleft': isNoneFields,
            },
          ]"
        >
          <el-form-item
            :style="{ transform: `translateX(${searchBtnOffset}px)` }"
            class="lv__filterbar-action-submit"
          >
            <slot name="prepend-submit" />
            <el-button
              v-if="isShowSearchButton"
              v-bind="searchButton"
              @click="handleFilterSearch"
            >
              {{ searchButton.text }}
            </el-button>
            <el-button
              v-if="isShowResetButton"
              v-bind="resetButton"
              @click="handleFilterReset"
            >
              {{ resetButton.text }}
            </el-button>
            <slot name="append-submit" />
          </el-form-item>
          <div class="lv__filterbar-action-ext">
            <slot name="prepend-more" />
            <el-button
              icon="el-icon-caret-top"
              type="primary"
              class="lv__filterbar-action-more"
              @click="toggleFilterbar"
            />
            <slot name="append-more" />
          </div>
        </div>

        <!-- 操作按钮区域 -->
        <FilterbarButtons
          v-if="isShowFilterButtons"
          :buttons="filterButtons"
          class="lv__filterbar-buttons"
        />

        <!-- 搜索栏控件区域 -->
        <FilterbarFields
          v-if="isShowFilterFields"
          ref="FilterbarFields"
          :fields="filterFields"
          class="lv__filterbar-fields"
        />
      </div>

      <div v-if="$slots['filterbar-right']" class="lv__filterbar-right">
        <slot name="filterbar-right" />
      </div>
    </div>

    <div v-if="$slots['filterbar-bottom']" class="lv__filterbar-bottom">
      <slot name="filterbar-bottom" />
    </div>
  </el-form>
</template>

<script lang="tsx">
import Vue, { PropType } from 'vue'
import { hasOwn } from '@/utils'
import FilterbarButtons from './FilterbarButtons.vue'
import FilterbarFields from './FilterbarFields.vue'
import storeProviderMixin from '@/mixins/storeProviderMixin'
import { FilterButton, FilterField } from '~/types'

export default Vue.extend({
  name: 'Filterbar',

  mixins: [storeProviderMixin],

  inheritAttrs: false,

  components: {
    FilterbarButtons,
    FilterbarFields,
  },

  props: {
    filterButtons: {
      type: Array as PropType<FilterButton[]>,
      default: /* istanbul ignore next */ () => [],
    },
    filterFields: {
      type: Array as PropType<FilterField[]>,
      default: /* istanbul ignore next */ () => [],
    },
    searchButton: {
      type: [Object, Boolean],
      default: () => ({
        text: '搜索',
        icon: 'el-icon-search',
        type: 'primary',
      }),
    },
    resetButton: {
      type: [Object, Boolean],
      default: () => ({
        text: '重置',
        icon: '',
        type: 'default',
      }),
    },
  },

  data() {
    return {
      isFold: true,
      isNoMore: true,
      topRightItemIndex: -1,
      actionOffsetLeft: 0,
      searchBtnOffset: 0,
    }
  },

  computed: {
    isNoneFields(): boolean {
      return this.filterFields.length === 0
    },
    isShowFilterButtons(): boolean {
      return this.filterButtons.length > 0
    },
    isShowFilterFields(): boolean {
      return this.filterFields.length > 0
    },
    isShowSearchButton(): boolean {
      return !!this.searchButton
    },
    isShowResetButton(): boolean {
      return !!this.resetButton
    },
    isShowFilterSubmit(): boolean {
      return !!(
        this.isShowSearchButton ||
        this.isShowResetButton ||
        this.$slots['prepend-filterbar-submit'] ||
        this.$slots['append-filterbar-submit']
      )
    },
    allFieldsVm(): Vue[] {
      try {
        return (this as any).$refs['FilterbarFields'].$refs['field'] || []
      } catch (error) {
        return []
      }
    },
  },

  watch: {
    isShowSearchButton: 'updateLayout',
    isShowResetButton: 'updateLayout',
  },

  methods: {
    handleSubmit() {
      this.lvStore.pressEnterSearch && this.handleFilterSearch()
    },

    handleFilterSearch() {
      this.$rootEmitProxy('filter-submit')
      this.lvStore.search()
    },

    handleFilterReset() {
      const filterModel = this.lvStore.filterModel
      const _resetField = (field: FilterField) => {
        const name = field.model
        if (name && hasOwn(filterModel, name)) {
          const value = filterModel[name]
          if (Array.isArray(value)) {
            this.$set(filterModel, name, [])
          } else {
            this.$set(filterModel, name, undefined)
          }
        }
      }
      this.filterFields.forEach((field) => {
        if (Array.isArray(field)) {
          field.forEach(_resetField)
        } else {
          _resetField(field)
        }
      })
      this.$rootEmitProxy('filter-reset')
    },

    toggleFilterbar() {
      this.isFold = !this.isFold
      this.$emit('fold-change', this.isFold)
    },

    updateLayout() {
      // updateTopRightItemIndex 影响 isNoMore 按钮显示，需计算后再执行按钮偏移量计算
      this.updateTopRightItemIndex()
      this.$nextTick().then(() => {
        this.updateActionOffset()
        this.updateBtnOffset()
      })
    },

    getAllFieldsDom() {
      try {
        return (this as any).$refs['FilterbarFields'].$refs['field'] || []
      } catch (error) {
        return []
      }
    },

    updateTopRightItemIndex() {
      let lastFilterIndex = -1
      const allFields = this.allFieldsVm
      if (allFields.length > 0) {
        const $action = this.$refs.action as Element
        let lastFilterTop = $action.getBoundingClientRect().top
        for (let i = 0; i < allFields.length; i++) {
          const curItemTop = allFields[i].$el.getBoundingClientRect().top
          if (lastFilterTop < curItemTop) {
            break
          }
          lastFilterIndex = i
        }
      }
      this.topRightItemIndex = lastFilterIndex
      this.isNoMore = this.topRightItemIndex === this.filterFields.length - 1
    },

    updateBtnOffset() {
      let offset = 0
      if (this.topRightItemIndex >= 0) {
        const lastItem = this.allFieldsVm[this.topRightItemIndex].$el
        const { left, width } = lastItem.getBoundingClientRect()
        offset = left + width - this.actionOffsetLeft + 10
        offset = Math.min(0, offset)
      }
      this.searchBtnOffset = Math.floor(offset)
    },

    updateActionOffset() {
      const $action = this.$refs.action as Element
      if ($action) {
        this.actionOffsetLeft = $action.getBoundingClientRect().left
      }
    },
  },
})
</script>

<style lang="less">
@filter-gap-size: 10px;

.lv__filterbar-top {
  margin-bottom: @filter-gap-size;
}
.lv__filterbar-bottom {
  margin-bottom: @filter-gap-size;
}
.lv__filterbar-left {
  margin-right: @filter-gap-size;
  margin-bottom: @filter-gap-size;
}
.lv__filterbar-right {
  margin-left: @filter-gap-size;
  margin-bottom: @filter-gap-size;
}
.lv__filterbar-main {
  display: flex;
}

.lv__filterbar {
  .el-button {
    [class*=' el-icon-'],
    [class^='el-icon-'] {
      width: 1em;
    }

    height: 32px;
    padding-top: 0;
    padding-bottom: 0;
    line-height: 1;
    vertical-align: top;
  }
  .el-input-group__append,
  .el-input-group__prepend {
    padding: 0 10px;
  }
  .el-form-item__label {
    padding: 0 0 0 2px;
    margin-right: -4px;
  }
  .el-form-item {
    margin: 0;
  }
}

.lv__filterbar {
  &-buttons {
    float: left;
    margin-right: 0;

    .el-button + .el-dropdown,
    .el-dropdown + .el-button,
    .el-dropdown + .el-dropdown,
    .el-form-item__content > * {
      display: inline-block;
      margin-right: @filter-gap-size;
      margin-left: 0;
    }
  }

  &-action {
    display: flex;
    float: right;
    margin: 0;
    margin-bottom: 10px;

    &-submit .el-form-item__content,
    &-ext {
      display: flex;
      & > * {
        display: inline-block;
        transition: inherit;
      }
      & > *:not(:nth-child(1)) {
        margin-left: 10px;
      }
    }

    &-ext {
      // display: flex;
      float: right;
      margin-left: 10px;
    }

    &-more {
      width: 40px;
      padding: 0;
      line-height: 30px;
      transition: none;
    }

    &--nomore {
      .lv__filterbar-action-more {
        display: none;
      }
    }

    &--onleft {
      float: none;
      display: inline-block;
    }
  }
}

.lv__filterbar-inner {
  flex: 1;
  padding-top: @filter-gap-size;
  margin-top: -@filter-gap-size;

  &::after {
    display: table;
    clear: both;
    content: '';
  }

  &--fold {
    box-sizing: content-box;
    height: 32px;
    overflow: hidden;
    margin-bottom: @filter-gap-size;

    .lv__filterbar-action-more {
      transform: rotate(180deg);
    }
  }
}
</style>
