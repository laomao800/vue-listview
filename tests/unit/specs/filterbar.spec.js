import { mount } from '@vue/test-utils'
import Filterbar from '@/listview/components/filterbar'
import {
  filterButtonsNormal,
  filterButtonsDropdown,
  filterFields
} from './props.js'

describe('filterbar - filterButtons', () => {
  it('普通按钮渲染', () => {
    const filterButtons = filterButtonsNormal
    const wrapper = mount(Filterbar, {
      propsData: { filterButtons }
    })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.findAll('.filterbar__buttons .el-button')).toHaveLength(
      filterButtons.length
    )
  })

  it('带下拉按钮渲染', () => {
    const filterButtons = filterButtonsDropdown
    const wrapper = mount(Filterbar, {
      propsData: { filterButtons }
    })
    // el-dropdown 会生成随机 id ，先删除后再测试 snapshots
    const randomAttrReg = /aria-controls="dropdown-menu-\d+"|id="dropdown-menu-\d+"/g
    let wrapperHtml = wrapper.html()
    wrapperHtml = wrapperHtml.replace(randomAttrReg, '')
    expect(wrapperHtml).toMatchSnapshot()
    expect(wrapper.findAll('.filterbar__buttons .el-dropdown')).toHaveLength(
      filterButtons.length
    )
  })

  it('JSX', function() {
    const jsxWrapper = {
      components: { Filterbar },
      render(h) {
        return (
          <filterbar
            filterButtons={[<button>button</button>, <strong>strong</strong>]}
          />
        )
      }
    }
    const wrapper = mount(jsxWrapper)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('按钮点击事件', () => {
    const spy = jest.fn()
    const wrapper = mount(Filterbar, {
      propsData: {
        filterButtons: [
          {
            text: 'default',
            click() {
              spy('default')
            }
          },
          {
            text: 'dropdown1',
            click() {
              spy('dropdown1')
            },
            children: [
              {
                text: 'sub1',
                click() {
                  spy('sub1')
                }
              }
            ]
          },
          {
            text: 'dropdown2',
            click() {
              spy('dropdown2')
            },
            splitButton: true,
            children: [
              {
                text: 'sub2',
                click() {
                  spy('sub2')
                }
              }
            ]
          }
        ]
      }
    })
    wrapper.find('.filterbar__buttons .el-button--default').trigger('click')
    const dropdown = wrapper.findAll('.filterbar__buttons .el-dropdown')
    dropdown
      .at(0)
      .find('.el-button')
      .trigger('click')
    dropdown
      .at(0)
      .find('.el-dropdown-menu .el-dropdown-menu__item')
      .trigger('click')
    dropdown
      .at(1)
      .find('.el-button')
      .trigger('click')
    dropdown
      .at(1)
      .find('.el-dropdown-menu .el-dropdown-menu__item')
      .trigger('click')

    expect(spy.mock.calls.length).toBe(5)
    expect(spy.mock.calls[0][0]).toBe('default')
    expect(spy.mock.calls[1][0]).toBe('dropdown1')
    expect(spy.mock.calls[2][0]).toBe('sub1')
    expect(spy.mock.calls[3][0]).toBe('dropdown2')
    expect(spy.mock.calls[4][0]).toBe('sub2')
  })
})

describe('filterbar - filterFields', () => {
  describe('FilterForm', () => {
    it('渲染所有有效的内置字段类型', () => {
      const wrapper = mount(Filterbar, {
        propsData: {
          filterFields: [
            {
              type: 'unknow',
              label: '无效类型'
            },
            ...filterFields
          ]
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
      expect(wrapper.findAll('.filterbar__field')).toHaveLength(
        filterFields.length
      )
    })

    it('指定 width', () => {
      const WIDTH = 280
      const cloneFields = JSON.parse(JSON.stringify(filterFields))
      Object.keys(cloneFields).forEach(key => {
        cloneFields[key].width = WIDTH
      })
      const wrapper = mount(Filterbar, {
        propsData: {
          filterFields: cloneFields
        }
      })
      const $fields = wrapper.findAll(
        '.filterbar__field > .el-form-item > .el-form-item__content > div'
      )
      $fields.wrappers.forEach(field => {
        expect(field.element.style.width).toBe(`${WIDTH}px`)
      })
    })

    it('JSX', function() {
      const jsxWrapper = {
        components: { Filterbar },
        render(h) {
          return (
            <filterbar
              filterFields={[<button>button</button>, <strong>strong</strong>]}
            />
          )
        }
      }
      const wrapper = mount(jsxWrapper)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})

describe('filterbar - filterbarFold', () => {
  const wrapper = mount(Filterbar, {
    propsData: {
      filterbarFold: true
    }
  })

  it('toggleFilterbar', () => {
    const foldClassName = 'list-view__filterbar--fold'
    const $toggler = wrapper.find('.filterbar__submit-more')
    const $filterbar = wrapper.find('.list-view__filterbar').element
    expect($filterbar.classList.contains(foldClassName)).toBe(true)
    wrapper.setProps({ filterbarFold: false })
    expect($filterbar.classList.contains(foldClassName)).toBe(false)
    $toggler.trigger('click')
    expect($filterbar.classList.contains(foldClassName)).toBe(true)
  })
})

describe('filterbar - SearchButton & ResetButton', () => {
  it('SearchButton', () => {
    const wrapper = mount(Filterbar, {
      propsData: { showFilterSearch: false }
    })
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.setProps({
      showFilterSearch: true
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('ResetButton', () => {
    const wrapper = mount(Filterbar, {
      propsData: { showFilterReset: false }
    })
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.setProps({
      showFilterReset: true
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Slots', () => {
    const slotWrapper = {
      components: { Filterbar },
      render(h) {
        return (
          <filterbar>
            <div slot="prepend-filterbar-submit">prepend-filterbar-submit</div>
            <div slot="append-filterbar-submit">append-filterbar-submit</div>
          </filterbar>
        )
      }
    }
    const wrapper = mount(slotWrapper)
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('Events', () => {
    const wrapper = mount(Filterbar)
    const searchBtn = wrapper.findAll('.filterbar__submit-btn button').at(0)
    const resetBtn = wrapper.findAll('.filterbar__submit-btn button').at(1)

    it('Search button', () => {
      searchBtn.trigger('click')
      expect(wrapper.emitted('filter-submit')).toBeTruthy()
      expect(wrapper.emitted('filter-submit').length).toBe(1)
    })

    it('Reset button', () => {
      resetBtn.trigger('click')
      expect(wrapper.emitted('filter-reset')).toBeTruthy()
      expect(wrapper.emitted('filter-reset').length).toBe(1)
    })

    it('Reset model', async () => {
      const wrapper = mount(Filterbar, {
        propsData: {
          filterFields: [
            { type: 'text', model: 'text' },
            { type: 'number', model: 'number' },
            { type: 'multipleSelect', model: 'multipleSelect' }
          ],
          filterModel: {
            text: 'mock text',
            number: 100,
            multipleSelect: [1, 2, 3],
            hiddenParam: 'hidden'
          }
        }
      })
      const resetBtn = wrapper.findAll('.filterbar__submit-btn button').at(1)
      resetBtn.trigger('click')
      expect(wrapper.vm.filterModel).toEqual({
        text: undefined,
        number: undefined,
        multipleSelect: [],
        hiddenParam: 'hidden'
      })
    })
  })
})
