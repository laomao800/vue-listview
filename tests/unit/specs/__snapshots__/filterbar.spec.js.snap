// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterbar - SearchButton & ResetButton ResetButton 1`] = `
<div class="list-view__filterbar list-view__filterbar--fold">
  <form class="el-form el-form--inline">
    <!---->
    <div class="filterbar__submit filterbar__submit--nomore filterbar__submit--onleft">
      <div class="filterbar__submit-btn" style="transform: translateX(0px);">
        <div class="el-form-item el-form-item--small">
          <!---->
          <div class="el-form-item__content"> <button type="button" class="el-button el-button--primary el-button--small">
              <!----><i class="el-icon-search"></i><span>
            搜索
          </span></button>
            <!---->
            <!---->
          </div>
        </div>
      </div> <button type="button" class="el-button filterbar__submit-more el-button--primary">
        <!----><i class="el-icon-caret-bottom"></i>
        <!----></button>
    </div>
    <!---->
  </form>
</div>
`;

exports[`filterbar - SearchButton & ResetButton ResetButton 2`] = `
<div class="list-view__filterbar list-view__filterbar--fold">
  <form class="el-form el-form--inline">
    <!---->
    <div class="filterbar__submit filterbar__submit--nomore filterbar__submit--onleft">
      <div class="filterbar__submit-btn" style="transform: translateX(0px);">
        <div class="el-form-item el-form-item--small">
          <!---->
          <div class="el-form-item__content"> <button type="button" class="el-button el-button--primary el-button--small">
              <!----><i class="el-icon-search"></i><span>
            搜索
          </span></button> <button type="button" class="el-button el-button--default el-button--small">
              <!---->
              <!----><span>
            重置
          </span></button>
            <!---->
          </div>
        </div>
      </div> <button type="button" class="el-button filterbar__submit-more el-button--primary">
        <!----><i class="el-icon-caret-bottom"></i>
        <!----></button>
    </div>
    <!---->
  </form>
</div>
`;

exports[`filterbar - SearchButton & ResetButton SearchButton 1`] = `
<div class="list-view__filterbar list-view__filterbar--fold">
  <form class="el-form el-form--inline">
    <!---->
    <div class="filterbar__submit filterbar__submit--nomore filterbar__submit--onleft">
      <div class="filterbar__submit-btn" style="transform: translateX(0px);">
        <div class="el-form-item el-form-item--small">
          <!---->
          <div class="el-form-item__content">
            <!----> <button type="button" class="el-button el-button--default el-button--small">
              <!---->
              <!----><span>
            重置
          </span></button>
            <!---->
          </div>
        </div>
      </div> <button type="button" class="el-button filterbar__submit-more el-button--primary">
        <!----><i class="el-icon-caret-bottom"></i>
        <!----></button>
    </div>
    <!---->
  </form>
</div>
`;

exports[`filterbar - SearchButton & ResetButton SearchButton 2`] = `
<div class="list-view__filterbar list-view__filterbar--fold">
  <form class="el-form el-form--inline">
    <!---->
    <div class="filterbar__submit filterbar__submit--nomore filterbar__submit--onleft">
      <div class="filterbar__submit-btn" style="transform: translateX(0px);">
        <div class="el-form-item el-form-item--small">
          <!---->
          <div class="el-form-item__content"> <button type="button" class="el-button el-button--primary el-button--small">
              <!----><i class="el-icon-search"></i><span>
            搜索
          </span></button> <button type="button" class="el-button el-button--default el-button--small">
              <!---->
              <!----><span>
            重置
          </span></button>
            <!---->
          </div>
        </div>
      </div> <button type="button" class="el-button filterbar__submit-more el-button--primary">
        <!----><i class="el-icon-caret-bottom"></i>
        <!----></button>
    </div>
    <!---->
  </form>
</div>
`;

exports[`filterbar - SearchButton & ResetButton Slots 1`] = `
<div class="list-view__filterbar list-view__filterbar--fold">
  <form class="el-form el-form--inline">
    <!---->
    <div class="filterbar__submit filterbar__submit--nomore filterbar__submit--onleft">
      <div class="filterbar__submit-btn" style="transform: translateX(0px);">
        <div class="el-form-item el-form-item--small">
          <!---->
          <div class="el-form-item__content">
            <div>prepend-filterbar-submit</div> <button type="button" class="el-button el-button--primary el-button--small">
              <!----><i class="el-icon-search"></i><span>
            搜索
          </span></button> <button type="button" class="el-button el-button--default el-button--small">
              <!---->
              <!----><span>
            重置
          </span></button>
            <div>append-filterbar-submit</div>
            <!---->
          </div>
        </div>
      </div> <button type="button" class="el-button filterbar__submit-more el-button--primary">
        <!----><i class="el-icon-caret-bottom"></i>
        <!----></button>
    </div>
    <!---->
  </form>
</div>
`;

exports[`filterbar - filterButtons JSX 1`] = `
<div class="list-view__filterbar list-view__filterbar--fold">
  <form class="el-form el-form--inline">
    <div class="filterbar__buttons">
      <div class="el-form-item el-form-item--small">
        <!---->
        <div class="el-form-item__content"><button>button</button><strong>strong</strong>
          <!---->
        </div>
      </div>
    </div>
    <div class="filterbar__submit filterbar__submit--nomore filterbar__submit--onleft">
      <div class="filterbar__submit-btn" style="transform: translateX(0px);">
        <div class="el-form-item el-form-item--small">
          <!---->
          <div class="el-form-item__content"> <button type="button" class="el-button el-button--primary el-button--small">
              <!----><i class="el-icon-search"></i><span>
            搜索
          </span></button> <button type="button" class="el-button el-button--default el-button--small">
              <!---->
              <!----><span>
            重置
          </span></button>
            <!---->
          </div>
        </div>
      </div> <button type="button" class="el-button filterbar__submit-more el-button--primary">
        <!----><i class="el-icon-caret-bottom"></i>
        <!----></button>
    </div>
    <!---->
  </form>
</div>
`;

exports[`filterbar - filterButtons 带下拉按钮渲染 1`] = `
<div class="list-view__filterbar list-view__filterbar--fold">
  <form class="el-form el-form--inline">
    <div class="filterbar__buttons">
      <div class="el-form-item el-form-item--small">
        <!---->
        <div class="el-form-item__content">
          <div class="el-dropdown">
            <div class="el-button-group"><button type="button" class="el-button el-button--primary el-button--small">
                <!---->
                <!----><span><i class="el-icon-circle-plus-outline"></i>
              下拉按钮1
             </span></button><button type="button" class="el-button el-button--primary el-button--small el-dropdown__caret-button" aria-haspopup="list">
                <!---->
                <!----><span><i class="el-dropdown__icon el-icon-arrow-down"></i></span></button></div>
            <ul class="el-dropdown-menu el-popper" style="display: none;" name="el-zoom-in-top">
              <li tabindex="-1" class="el-dropdown-menu__item">
                <!---->
                菜单1
              </li>
              <li tabindex="-1" class="el-dropdown-menu__item">
                <!---->
                菜单2
              </li>
            </ul>
          </div>
          <div class="el-dropdown"><button type="button" class="el-button el-button--primary el-button--small el-dropdown-selfdefine" aria-haspopup="list" role="button" tabindex="0">
              <!---->
              <!----><span>
                下拉按钮2<i class="el-icon-arrow-down el-icon--right"></i></span></button>
            <ul class="el-dropdown-menu el-popper" style="display: none;" name="el-zoom-in-top">
              <li tabindex="-1" class="el-dropdown-menu__item">
                <!---->
                菜单1
              </li>
              <li tabindex="-1" class="el-dropdown-menu__item">
                <!---->
                菜单2
              </li>
            </ul>
          </div>
          <!---->
        </div>
      </div>
    </div>
    <div class="filterbar__submit filterbar__submit--nomore filterbar__submit--onleft">
      <div class="filterbar__submit-btn" style="transform: translateX(0px);">
        <div class="el-form-item el-form-item--small">
          <!---->
          <div class="el-form-item__content"> <button type="button" class="el-button el-button--primary el-button--small">
              <!----><i class="el-icon-search"></i><span>
            搜索
          </span></button> <button type="button" class="el-button el-button--default el-button--small">
              <!---->
              <!----><span>
            重置
          </span></button>
            <!---->
          </div>
        </div>
      </div> <button type="button" class="el-button filterbar__submit-more el-button--primary">
        <!----><i class="el-icon-caret-bottom"></i>
        <!----></button>
    </div>
    <!---->
  </form>
</div>
`;

exports[`filterbar - filterButtons 普通按钮渲染 1`] = `
<div class="list-view__filterbar list-view__filterbar--fold">
  <form class="el-form el-form--inline">
    <div class="filterbar__buttons">
      <div class="el-form-item el-form-item--small">
        <!---->
        <div class="el-form-item__content"><button type="button" class="el-button el-button--default el-button--small">
            <!---->
            <!----><span>
            default
          </span></button><button type="button" class="el-button el-button--text el-button--small">
            <!---->
            <!----><span>
            text
          </span></button><button type="button" class="el-button el-button--primary el-button--small">
            <!----><i class="el-icon-edit"></i><span>
            primary
          </span></button><button type="button" class="el-button el-button--success el-button--small">
            <!----><i class="el-icon-check"></i><span>
            success
          </span></button><button type="button" class="el-button el-button--info el-button--small">
            <!----><i class="el-icon-message"></i><span>
            info
          </span></button><button type="button" class="el-button el-button--warning el-button--small">
            <!----><i class="el-icon-star-off"></i><span>
            warning
          </span></button><button type="button" class="el-button el-button--danger el-button--small">
            <!----><i class="el-icon-delete"></i><span>
            danger
          </span></button>
          <!---->
        </div>
      </div>
    </div>
    <div class="filterbar__submit filterbar__submit--nomore filterbar__submit--onleft">
      <div class="filterbar__submit-btn" style="transform: translateX(0px);">
        <div class="el-form-item el-form-item--small">
          <!---->
          <div class="el-form-item__content"> <button type="button" class="el-button el-button--primary el-button--small">
              <!----><i class="el-icon-search"></i><span>
            搜索
          </span></button> <button type="button" class="el-button el-button--default el-button--small">
              <!---->
              <!----><span>
            重置
          </span></button>
            <!---->
          </div>
        </div>
      </div> <button type="button" class="el-button filterbar__submit-more el-button--primary">
        <!----><i class="el-icon-caret-bottom"></i>
        <!----></button>
    </div>
    <!---->
  </form>
</div>
`;

exports[`filterbar - filterFields FilterForm JSX 1`] = `
<div class="list-view__filterbar list-view__filterbar--fold">
  <form class="el-form el-form--inline">
    <!---->
    <div class="filterbar__submit filterbar__submit--nomore">
      <div class="filterbar__submit-btn" style="transform: translateX(0px);">
        <div class="el-form-item el-form-item--small">
          <!---->
          <div class="el-form-item__content"> <button type="button" class="el-button el-button--primary el-button--small">
              <!----><i class="el-icon-search"></i><span>
            搜索
          </span></button> <button type="button" class="el-button el-button--default el-button--small">
              <!---->
              <!----><span>
            重置
          </span></button>
            <!---->
          </div>
        </div>
      </div> <button type="button" class="el-button filterbar__submit-more el-button--primary">
        <!----><i class="el-icon-caret-bottom"></i>
        <!----></button>
    </div>
    <div class="filterbar__form">
      <div class="filterbar__field">
        <!----> <button>button</button></div>
      <div class="filterbar__field">
        <!----> <strong>strong</strong></div>
    </div>
  </form>
</div>
`;

exports[`filterbar - filterFields FilterForm 渲染所有有效的内置字段类型 1`] = `
<div>
  <div class="filterbar__field">
    <!---->
    <elformitem-stub inlinemessage="" showmessage="true">
      <fieldtext-stub model="[object Object]" field="[object Object]"></fieldtext-stub>
    </elformitem-stub>
  </div>
  <div class="filterbar__field">
    <!---->
    <elformitem-stub inlinemessage="" showmessage="true">
      <fieldtext-stub model="[object Object]" field="[object Object]"></fieldtext-stub>
    </elformitem-stub>
  </div>
  <div class="filterbar__field">
    <!---->
    <elformitem-stub inlinemessage="" showmessage="true">
      <fieldselect-stub model="[object Object]" field="[object Object]"></fieldselect-stub>
    </elformitem-stub>
  </div>
  <div class="filterbar__field">
    <!---->
    <elformitem-stub inlinemessage="" showmessage="true">
      <fieldmultipleselect-stub model="[object Object]" field="[object Object]"></fieldmultipleselect-stub>
    </elformitem-stub>
  </div>
  <div class="filterbar__field">
    <!---->
    <elformitem-stub inlinemessage="" showmessage="true">
      <fielddate-stub model="[object Object]" field="[object Object]"></fielddate-stub>
    </elformitem-stub>
  </div>
  <div class="filterbar__field">
    <!---->
    <elformitem-stub inlinemessage="" showmessage="true">
      <fielddaterange-stub model="[object Object]" field="[object Object]"></fielddaterange-stub>
    </elformitem-stub>
  </div>
  <div class="filterbar__field">
    <!---->
    <elformitem-stub inlinemessage="" showmessage="true">
      <fieldtimeselect-stub model="[object Object]" field="[object Object]"></fieldtimeselect-stub>
    </elformitem-stub>
  </div>
  <div class="filterbar__field">
    <!---->
    <elformitem-stub inlinemessage="" showmessage="true">
      <fieldtimepicker-stub model="[object Object]" field="[object Object]"></fieldtimepicker-stub>
    </elformitem-stub>
  </div>
  <div class="filterbar__field">
    <!---->
    <elformitem-stub inlinemessage="" showmessage="true">
      <fieldtimepickerrange-stub model="[object Object]" field="[object Object]"></fieldtimepickerrange-stub>
    </elformitem-stub>
  </div>
  <div class="filterbar__field">
    <!---->
    <elformitem-stub inlinemessage="" showmessage="true">
      <fielddatetime-stub model="[object Object]" field="[object Object]"></fielddatetime-stub>
    </elformitem-stub>
  </div>
  <div class="filterbar__field">
    <!---->
    <elformitem-stub inlinemessage="" showmessage="true">
      <fielddatetimerange-stub model="[object Object]" field="[object Object]"></fielddatetimerange-stub>
    </elformitem-stub>
  </div>
  <div class="filterbar__field">
    <!---->
    <elformitem-stub inlinemessage="" showmessage="true">
      <fieldcascader-stub model="[object Object]" field="[object Object]"></fieldcascader-stub>
    </elformitem-stub>
  </div>
</div>
`;
